File: Android/app/src/main/java/net.neczpal.interstellarwar.common/game/Road.java
package net.neczpal.interstellarwar.common.game;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public class Road implements Serializable {
    private static final long serialVersionUID = 1683452581122892189L;

    private RoadKey mKey;
    private Planet mFrom, mTo;
    private List<SpaceShip> mSpaceShips;

    /**
     * Erstellt ein Weg zwischen zwei Planet
     *
     * @param from der eine Planet
     * @param to   der andere Planet
     */
    public Road (RoadKey key, Planet from, Planet to) {
        this.mKey = key;
        this.mFrom = from;
        this.mTo = to;
        this.mSpaceShips = new ArrayList<> ();
    }

    public RoadKey getKey () {
        return mKey;
    }

    // GETTERS

    public void addSpaceship (SpaceShip spaceShip) {
        if (!mSpaceShips.contains (spaceShip))
            mSpaceShips.add (spaceShip);
    }

    public void removeSpaceships (SpaceShip spaceShips) {
        mSpaceShips.remove (spaceShips);
    }

    public void removeSpaceships (List<SpaceShip> spaceShips) {
        mSpaceShips.removeAll (spaceShips);
    }

    public List<SpaceShip> getSpaceShips () {
        return mSpaceShips;
    }

    public Planet getFrom () {
        return mFrom;
    }

    public Planet getTo () {
        return mTo;
    }

    @Override
    public boolean equals (Object o) {
        if (this == o) return true;
        if (o == null || getClass () != o.getClass ()) return false;
        Road road = (Road) o;
        return ((Objects.equals (mFrom, road.mFrom) && Objects.equals (mTo, road.mTo)) ||
                Objects.equals (mFrom, road.mTo) && Objects.equals (mTo, road.mFrom));
    }

    static class RoadKey {
        public int from;
        public int to;

        public RoadKey (int from, int to) {
            this.from = from;
            this.to = to;
        }

        @Override
        public boolean equals (Object o) {
            if (this == o) return true;
            if (o == null || getClass () != o.getClass ()) return false;
            RoadKey roadKey = (RoadKey) o;
            return (from == roadKey.from &&
                    to == roadKey.to) ||
                    (from == roadKey.to &&
                            to == roadKey.from);
        }

        @Override
        public int hashCode () {
            return Objects.hash (from, to) + Objects.hash (to, from);
        }
    }

    @Override
    public int hashCode () {
        return Objects.hash (mFrom, mTo);
    }
}

File: iOS/InterstellarWar/InterstellarWar/Model/game/Road.swift
//
//  Road.swift
//  InterstellarWar
//
//  Created by Neczpál Ábel on 2020. 01. 30..
//  Copyright © 2020. Neczpál Ábel. All rights reserved.
//

import SpriteKit

public class RoadKey : Hashable {
    public func hash(into hasher: inout Hasher) {
        hasher.combine((from + to) * ((from+1) * (to+1)))
    }
    
    public static func == (lhs: RoadKey, rhs: RoadKey) -> Bool {
        return (lhs.from == rhs.from && lhs.to == rhs.to) ||
            (lhs.to == rhs.from && lhs.from == rhs.to)
    }
    
    public var from : Int
    public var to : Int
    
    init (_ from: Int, _ to: Int) {
        self.from = from
        self.to = to
    }
}

public class Road {
    
    private var mKey : RoadKey
    
    private var mFrom : Planet, mTo : Planet
    
    private var mSpaceShips : [SpaceShip]
    
    private var mNode : SKShapeNode

    
    init(key: RoadKey, from : Planet, to : Planet) {
        mKey = key
        mFrom = from
        mTo = to
        mSpaceShips = [SpaceShip] ()
        
        
        let path = CGMutablePath()
        path.move(to: CGPoint(x: from.getX(), y: from.getY()))
        path.addLine(to: CGPoint(x: to.getX(), y: to.getY()))
        
        mNode = SKShapeNode(path: path)
    }

    // GETTERS

    public func getFrom () -> Planet {
        return mFrom
    }

    public func getTo () -> Planet {
        return mTo
    }
    
    public func addSpaceship (spaceShip: SpaceShip) {
        if (!mSpaceShips.contains (spaceShip)){
            mSpaceShips.append (spaceShip);
        }
    }

//    public func removeSpaceship (spaceShip: SpaceShip) {
//        mSpaceShips.remove
//    }
//
//    public func removeSpaceships (spaceShips: SpaceShip) {
//        mSpaceShips.removeAl
//    }

    public func getSpaceShips () -> [SpaceShip] {
        return mSpaceShips;
    }
    
    public func getNode () -> SKShapeNode {
        return mNode
    }


}
